api/src/main/java/org/openmrs/module/locationbasedaccess/PatientSearchAdviser.java

```diff
@@ -0,0 +1,89 @@
+import org.apache.commons.logging.LogFactory;
+import org.openmrs.Location;
+import org.openmrs.Patient;
+import org.openmrs.PersonAttribute;
+import org.openmrs.PersonAttributeType;
+import org.openmrs.api.context.Context;
+import org.springframework.aop.Advisor;
+import org.springframework.aop.support.StaticMethodMatcherPointcutAdvisor;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+public class PatientSearchAdviser extends StaticMethodMatcherPointcutAdvisor implements Advisor {
+
+    private static final Log log = LogFactory.getLog(PatientSearchAdviser.class);
+    public static final  String LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME = "locationbasedaccess.locationAttributeUuid";
+
+    @Override
+    public boolean matches(Method method, Class targetClass) {
+        if (method.getName().equals("getPatients")) {
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public Advice getAdvice() {
+        return new PatientSearchAdvise();
+    }
+
+    private class PatientSearchAdvise implements MethodInterceptor {
+
+        public Object invoke(MethodInvocation invocation) throws Throwable {
+            Integer sessionLocationId = Context.getUserContext().getLocationId();
+            String locationAttributeUuid = Context.getAdministrationService().getGlobalProperty(LOCATION_ATTRIBUTE_GLOBAL_PROPERTY_NAME);
+            List<Patient> patientList = (List<Patient>) invocation.proceed();
+            if (StringUtils.isNotBlank(locationAttributeUuid)) {
+                final PersonAttributeType personAttributeType = Context.getPersonService().getPersonAttributeTypeByUuid(locationAttributeUuid);
+                if (sessionLocationId != null) {
+                    Location sessionLocation = Context.getLocationService().getLocation(sessionLocationId);
+
+                    for (Iterator<Patient> iterator = patientList.iterator(); iterator.hasNext(); ) {
+                        Patient patient = iterator.next();
+                        if (!isPatientConstainsLocation(sessionLocation, patient.getAttribute(personAttributeType))) {
+                            iterator.remove();
+                        }
+                    }
+                } else {
+                    // If the sessionLocationId is null, then return a empty list
+                    log.debug("Search Patient : Null Session Location in the UserContext");
+                    return new ArrayList<Patient>();
+                }
+            }
+            return patientList;
+        }
+
+        public Boolean isPatientConstainsLocation(Location location, PersonAttribute personAttribute ) {
+            if (personAttribute == null) {
+                log.debug("Patient AccessLocation attribute missing for Patient");
+                return false;
+            }
+            else {
+                if(!personAttribute.getValue().equals(location.getUuid())) {
```

Comment:
Is it efficient to call location.getUuid() for each and every patient?

Reply:
We can improve it by replacing UUID.

Reply:
Changed